src/engine/platform/sms.cpp

```diff
@@ -396,45 +444,88 @@ void DivPlatformSMS::notifyInsDeletion(void* ins) {
 }
 
 void DivPlatformSMS::poke(unsigned int addr, unsigned short val) {
-  rWrite(val);
+  rWrite(addr,val);
 }
 
 void DivPlatformSMS::poke(std::vector<DivRegWrite>& wlist) {
-  for (DivRegWrite& i: wlist) rWrite(i.val);
+  for (DivRegWrite& i: wlist) rWrite(i.addr,i.val);
 }
 
 void DivPlatformSMS::setFlags(unsigned int flags) {
-  if ((flags&3)==3) {
-    chipClock=COLOR_NTSC/2.0;
-  } else if ((flags&3)==2) {
-    chipClock=4000000;
-  } else if ((flags&3)==1) {
-    chipClock=COLOR_PAL*4.0/5.0;
-  } else {
-    chipClock=COLOR_NTSC;
+  switch (flags&0xff03) {
+    default:
+    case 0x0000:
+      chipClock=COLOR_NTSC;
+      break;
+    case 0x0001:
+      chipClock=COLOR_PAL*4.0/5.0;
+      break;
+    case 0x0002:
+      chipClock=4000000;
+      break;
+    case 0x0003:
+      chipClock=COLOR_NTSC/2.0;
+      break;
+    case 0x0100:
+      chipClock=3000000;
+      break;
+    case 0x0101:
+      chipClock=2000000;
+      break;
   }
   resetPhase=!(flags&16);
-  
+  noiseDivider=64;
   if (sn!=NULL) delete sn;
-  switch ((flags>>2)&3) {
-    case 1: // TI
-      sn=new sn76496_base_device(0x4000, 0x4000, 0x01, 0x02, true, 1, false, true);
+  switch (flags&0xcc) {
+    default: // Sega
+    case 0x00:
+      sn=new segapsg_device();
+      isRealSN=false;
+      stereo=false;
+      break;
+    case 0x04: // TI SN76489
+      sn=new sn76489_device();
       isRealSN=true;
+      noiseDivider=60;
+      stereo=false;
       break;
-    case 2: // TI+Atari
-      sn=new sn76496_base_device(0x4000, 0x0f35, 0x01, 0x02, true, 1, false, true);
+    case 0x08: // TI+Atari
+      sn=new sn76496_base_device(0x4000, 0x0f35, 0x01, 0x02, true, false, 8, false, true);
       isRealSN=true;
+      noiseDivider=60;
+      stereo=false;
       break;
-    case 3: // Game Gear (not fully emulated yet!)
-      sn=new sn76496_base_device(0x8000, 0x8000, 0x01, 0x08, false, 1, false, false);
+    case 0x0c: // Game Gear (not fully emulated yet!)
+      sn=new gamegear_device();
       isRealSN=false;
+      stereo=true;
       break;
-    default: // Sega
-      sn=new sn76496_base_device(0x8000, 0x8000, 0x01, 0x08, false, 1, false, false);
+    case 0x40: // TI SN76489A
+      sn=new sn76489a_device();
+      isRealSN=false; // TODO
+      noiseDivider=68;
+      stereo=false;
+      break;
+    case 0x44: // TI SN76496
+      sn=new sn76496_device();
+      isRealSN=false; // TODO
+      noiseDivider=68;
+      stereo=false;
+      break;
+    case 0x48: // NCR 8496
+      sn=new ncr8496_device();
+      isRealSN=false;
+      noiseDivider=64;
+      stereo=false;
+      break;
+    case 0x4c: // Tandy PSSJ 3-voice sound
+      sn=new pssj3_device();
       isRealSN=false;
+      noiseDivider=64;
+      stereo=false;
       break;
   }
-  rate=chipClock/16;
+  rate=nuked?chipClock/16:chipClock/2;
```

Comment:
No, please no.
Running the chip at a faster rate will increase CPU load which is undesirable.